crates/tools/src/lib.rs

```diff
@@ -41,3 +54,57 @@ pub fn collect_tests(s: &str) -> Vec<(usize, Test)> {
     }
     res
 }
+
+
+pub fn update(path: &Path, contents: &str, verify: bool) -> Result<()> {
+    match fs::read_to_string(path) {
+        Ok(ref old_contents) if old_contents == contents => {
+            return Ok(());
+        }
+        _ => (),
+    }
+    if verify {
+        bail!("`{}` is not up-to-date", path.display());
+    }
+    eprintln!("updating {}", path.display());
+    fs::write(path, contents)?;
+    Ok(())
+}
+
+pub fn render_template(template: &str) -> Result<String> {
+    let grammar: ron::value::Value = {
+        let text = fs::read_to_string(format!("{}{}", Path::new(&std::env::var("CARGO_MANIFEST_DIR").unwrap()).parent().unwrap().to_str().unwrap(), "/ra_syntax/src/grammar.ron"))?;
```

Comment:
@matklad Here I have used `CARGO_MANIFEST_DIR` for locating the grammar.ron file. But it looks too complex to me.`CARGO_MANIFEST_DIR` returns `/home/mominul/src/rust-analyzer/crates/tools`, to remove the last path component I had to use `Path`'s `.parent()` method. Do you have any other way to do it in your mind or this is acceptable?

Thanks!

Reply:
I think that's OK, though I'd rather add an

```
fn project_root() -> PathBuf {
}
```

method, which does `.parent` stuff, so that this could look like `project_root().join("/crates/src/grammar.ron")`.

You also don't need to go via `format`, you can work on paths directly!

Reply:
Thanks! For the other paths, should I do them inline or make each a function to get the paths?

Reply:
I think inline is OK, but it probably make it so that the paths start from the repository root, so, somethng like

```
const SYNTAX_KINDS: &str = "crates/ra_syntax/src/syntax_kinds/generated.rs";
```